src/config.rs

```diff
@@ -14,8 +23,152 @@ impl Config {
+        Ok(Self {
+            config: Config::parse_from_toml(&workspace.join(".cargo/config.toml"))?,
+            workspace,
+        })
+    }
+
+    /// Search for `.cargo/config.toml` in any parent of the workspace root path.
+    /// Returns the directory which contains this path, not the path to the config file.
+    fn find_cargo_config_parent(workspace: impl AsRef<Path>) -> Result<Option<PathBuf>, Error> {
+        let workspace = dunce::canonicalize(workspace)?;
+        Ok(workspace
+            .ancestors()
+            .find(|dir| dir.join(".cargo/config.toml").is_file())
+            .map(|p| p.to_path_buf()))
+    }
+
+    /// Search for and open `.cargo/config.toml` in any parent of the workspace root path.
+    pub fn find_cargo_config_for_workspace(
+        workspace: impl AsRef<Path>,
+    ) -> Result<Option<Self>, Error> {
+        let config = Self::find_cargo_config_parent(workspace)?;
+        config.map(LocalizedConfig::new).transpose()
+    }
+
+    /// Read an environment variable from the `[env]` section in this `.cargo/config.toml`.
+    /// it is interpreted as path and canonicalized relative to [`Self::workspace`] if
+    /// [`EnvOption::Value::relative`] is set.
+    ///
+    /// Application environment variables (from [`std::env::var()`]) have [precedence]
+    /// unless [`EnvOption::Value::force`] is set. Note that this returns [`None`] if the
+    /// variable is not present in `[env]`, even if it may be available in [`std::env::var()`].
+    ///
+    /// [precedence]: https://doc.rust-lang.org/cargo/reference/config.html#env
+    pub fn resolve_env(&self, key: &str) -> io::Result<Option<Cow<'_, str>>> {
+        self.config
+            .env
+            .get(key)
+            .map(|v| v.resolve_value(key, &self.workspace))
+            .transpose()
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq)]
+#[serde(rename_all = "kebab-case")]
 pub struct Build {
-    #[serde(rename = "target-dir")]
     pub target_dir: Option<String>,
 }
+
+/// Serializable environment variable in cargo config, configurable as per
+/// <https://doc.rust-lang.org/cargo/reference/config.html#env>,
+#[derive(Clone, Debug, Deserialize, PartialEq)]
+#[serde(untagged, rename_all = "kebab-case")]
+pub enum EnvOption {
+    String(String),
+    Value {
+        value: String,
+        #[serde(default)]
+        force: bool,
+        #[serde(default)]
+        relative: bool,
+    },
+}
+
+impl EnvOption {
+    /// `config_parent` is the directory containing `.cargo/config.toml` where this was parsed from
+    pub fn resolve_value(
+        &self,
+        key: &str,
+        config_parent: impl AsRef<Path>,
+    ) -> io::Result<Cow<'_, str>> {
+        // Environment variables always have precedence, unless the extended format
+        // is used to set force=true
+        if !matches!(self, Self::Value { force: true, .. }) {
+            // TODO: Differentiate between invalid strings and key not found
```

Comment:
agree, this should be more like
```
if matches!(self, Self::Value { force: true, .. }) {
    return value from config
}
if let Ok(var) = std::env::var(key) {
    return var
}
if matches!(self, Self::Value { .. }) {
    return value from config
}
Err(couldn't resolve env variable) or None
```

Reply:
This `TODO` is specifically about returning more descriptive error messages, which is done now through a new `enum EnvError`.